### DISCUSSION TITLE
Constrain concentration variable

### DISCUSSION THREAD

**Discussion by souzanha** [idaholab/moose#17937] [Q&A Modules: Phase field] [answered]
Hi all, 
I get large fluctuations in my system with a regular Cahn-Hilliard + NeumannBC, and cannot uphold 0<= c <= 1 which gives me convergence problems, especially as I'm using a thermodynamic model with logarithmic terms. I've tried using polynomials, which allows me larger time-steps and faster convergence but then these fluctuations keeps happening until I reach c~10^2 and then it crashes. I use adaptivity (mesh is quite refined in the problematic areas). 

Now, I think I would like to try to constrain the concentration, i.e:

if (_c[qp] > 1) {
    _c[qp] = 1
}

Which I think I could include in any kernel, as long as I call for and initialize the concentration variable. However I'm quite new still with Moose, so I would value any suggestions.

Thanks in advance,
Souzan

**Comment by jiangwen84** [comment, marked as answer]
First thing I would check is your free energy. Looks like the system tries to find the solution/minimum outside 0-1 range. If that is the case, adding any constraint would not help you from getting a correct solution.

If you think your free energy is correct, then you can try to use a variational inequality solver (`snes_type = vinewtonrsls`) to set the bounds for your variable. One example can be found in test/tests/auxkernels/bounds/constant_bounds.i  

I am pinging @laagesen to provide you more suggestions.

**Reply by laagesen** [reply]
Hi Souzan,

I would start out by comparing your input file to one of the many Cahn-Hilliard examples in the MOOSE repository. For example this is a good one:

`moose/modules/phase_field/examples/cahn-hilliard/Parsed_SplitCH.i`

That can help you figure out if the problem lies with the governing equations in your input file. I would then try modifying the free energy in the example file to the polynomial form you want to use, then try the logarithmic form, to see if you can isolate whether the free energy is causing troubles. I would also try simulations without the NeumannBC at first, that may be causing solute to be added into the simulation domain and driving it away from two-phase coexistence. You mentioned that c~100 at the end, if it is that large everywhere in the domain that may be caused by the NeumannBC.

**Reply by souzanha** [reply]
Thank you @jiangwen84 and @laagesen for your suggestions. I will start again from the beginning to see if I can isolate the problem.

I started off with using the same defined energy expression, and changed it into mine. The polynomial function I use seem to work nicely, as does the logarithmic term. However, it is when I introduce the flux when it slows down and I start getting issues. Previously I had not included any periodic boundary condition, only the Neumann at one boundary. Now when I kept the periodic boundary block from 

moose/modules/phase_field/examples/cahn-hilliard/Parsed_SplitCH.i

... I can use larger time steps and the solution runs more smoothly. Is there a reason for that?

**Reply by laagesen** [reply]
If you are using a large value for the NeumannBC you may be adding a lot of solute to the system. This could be acting like a large source term and driving it far away from the concentration you start from. In such cases of a large effective source, it can make the problem more difficult to converge at each time step. In addition to the periodic BCs, I would also try running with no boundary conditions block, this will give you the natural BC in the finite element method, which in this case is a no-flux BC.

What is the physical system you are trying to simulate? We can probably help troubleshoot the NeumannBC but it would help understand the physics of your system and why you are wanting to add flux at a boundary.

**Reply by souzanha** [reply]
I see, I was also quite worried that this was the case. I've been using a NeumannBC, but for my real application I'd made my own boundary kernel that adds a flux depending on the chemical potential value at the boundary. 

I'm essentially following: https://pubs.rsc.org/en/content/articlehtml/2020/ta/c9ta11697d 

I may have defined my kernel incorrectly. The idea is to call for the value of the chemical potential at the boundary, use it in the Butler-Volmer equation and define an outgoing flux. 

This is how it looks like:

#include "ADButlerVolmerBC.h"

registerMooseObject("intercalationApp", ADButlerVolmerBC);

defineLegacyParams(ADButlerVolmerBC);

InputParameters
ADButlerVolmerBC::validParams()
{
    InputParameters params = ADIntegratedBC::validParams();
    params.addRequiredCoupledVar("w", "Coupled variable setting the gradient on the boundary.");
    params.addParam<Real>("Temperature", 298, "Value of temperature to use");
    params.addParam<Real>("current_density",1.0, "Applied Current Density");
    params.addParam<Real>("deltaPhi",100e-3, "Potential 100mV");
    params.addParam<Real>("V",43.8e-6, "Molar volume");
    params.addParam<Real>("F",96485, "Faradays constant");
    params.addParam<Real>("mueq",-3e5, "Chemical potential at phase equilibrium"); // J/m^3 
    params.addParam<Real>("R",8.315,"Gas constant");
    params.addParam<Real>("l_0",1e-9,"Length unit");
    params.addParam<Real>("e_0",1e-18,"Energy unit");
    params.addParam<Real>("D",1e-15,"Diffusion coefficient");
    return params;
}

ADButlerVolmerBC::ADButlerVolmerBC(const InputParameters & parameters)
        : ADIntegratedBC(parameters),
        _w(coupledValue("w")),
        _temp(getParam<Real>("Temperature")),
        _deltaPhi(getParam<Real>("deltaPhi")),
        _j0(getParam<Real>("current_density")),
        _V(getParam<Real>("V")),
        _F(getParam<Real>("F")),
        _mueq(getParam<Real>("mueq")),
        _R(getParam<Real>("R")),
        _lo(getParam<Real>("l_0")),
        _eo(getParam<Real>("e_0")),
        _D(getParam<Real>("D"))
{
}

ADReal
ADButlerVolmerBC::computeQpResidual()
{
    auto constant = (_mueq-(_F * _deltaPhi)) * ((_lo * _lo * _lo)/_eo); //scale dimension by m^3/J
    auto div = ((2 * _R * _temp)/_V) * ((_lo * _lo * _lo)/_eo); //scale the denominator in the exponent term
    auto exponent = (constant-_w[_qp])/div;
    auto flux = (((_lo * _lo * _lo * _lo)/_eo) * ((_R  * _temp)/(_F * _D)) * _j0  * (std::exp(exponent) - std::exp(-exponent)));

    return -_test[_i][_qp] * flux;
}

**Reply by dschwen** [reply]
What variable are you applying that BC to?

**Reply by dschwen** [reply]
Our default split formulation is a reverse split, which leads to counterintuitive variable/equation assignments. I think because of the coupledtimederivative you need to assign the kernel to the w variable and have c be the coupled variable.

**Reply by souzanha** [reply]
I use the kernels: CoupledTimeDerivative, SplitCHWRes and SplitCHParsed and have applied by BC to the chemical potential. 
However, I have also coupled to the chemical potential: 

[BCs]
  [./right]
     type = ADButlerVolmerBC
     variable = w
     w = w
     boundary = right
 [../]
[]

Now, I think I may not have fully understood the coupling correctly. What I know is that I need to call for the chemical potential at the nodes of the boundary, which is why I coupled to w. 

Running the model with the above kernels, it runs smoothly for awhile (total c keeps decreasing). But I get c_max ~  2.5 , c_min ~-1e-5, before it crashes.

**Reply by laagesen** [reply]
Can you be more specific about what you mean by "it crashes"? What is the error you are getting? You already mentioned that you don't have problems when you use periodic BCs. Do you get the same behavior if you use NeumannBC with a constant value?

**Reply by souzanha** [reply]
Thank you for the troubleshooting! I'm very glad to have some people to discuss this with. This is the error that I get:  

Time Step 284, time = 0.0900684, dt = 2e-14
 0 Nonlinear |R| = 2.759644e+11
      0 Linear |R| = 2.759644e+11
      1 Linear |R| = 1.169056e+01
 1 Nonlinear |R| = 8.527509e+08
      0 Linear |R| = 8.527509e+08
      1 Linear |R| = 2.677094e+00
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 1
 Solve Did NOT Converge!
Aborting as solve did not converge

*** ERROR ***
The following error occurred in the object "TimeStepper", of type "ConstantDT".

Solve failed and timestep already at or below dtmin, cannot continue!



I'm outputting the maximum and minimum concentrations and this happens when large fluctuations start occurring. 
c_max = 2.514514e+00 
c_,min = -5.985228e-01 

Regarding the periodic BCs, I tested it with moose/modules/phase_field/examples/cahn-hilliard/Parsed_SplitCH.i
when including a NeumannBC it converged easier and with a lower timestep. In the end, having a periodic BC is not an issue at the moment. But I would not want to apply it anywhere if I could. So I tried my model with the Butler-Volmer BC and added the Periodic BC, and I get much more far into the solution. I just let it run and I'm already much further ahead just by looking at the outputted time steps. Note that this is just by adding the Period BC. Could this have to do with any symmetry issues in FEM? I have not heard about periodic BC helping with convergence.

**Reply by laagesen** [reply]
From what you have said, the model runs fine with the standard periodic and standard Neumann BCs, but not when the Butler-Volmer BC that you created is included. So it seems the problem is most likely with the Butler-Volmer BC. So, I suggest focusing your troubleshooting efforts there. Can you create a simpler test case to debug the Butler-Volmer BC? It is always good to compare to an analytical solution if there is one available. If there is a 1D problem that has an analytical solution, that would be a great way to test and debug.

**Reply by souzanha** [reply]
Yes, I can reproduce the convergence issues by adding a high Neumann BC. So in turn, this means that my flux condition is large which creates fluctuations and numerical instabilities. I will have a look at it again. I think I'm connecting the variables correctly, but I may have to figure out how to deal with large fluxes somehow. 

However, in simple terms. The flux is dependent on the time-step right? Theoretically, even if we have a high flux the system should be able to adapt to it if the time-step is small enough. Am I correct in this? We have been able to set up MOOSE on our supercomputer and we will try decreasing the mesh and time-step in order to see if this may help.

**Reply by laagesen** [reply]
Yes flux is dependent on time step, i.e. as time step is lowered the total amount of solute passing the boundary is also lowered.

I'm wondering if you are using such a large flux that there is not enough time for diffusion to transport it in to the remainder of the system to any significant degree. In that case the system could be driven far away from equilibrium and have convergence problems. I'd suggest the following. Set up a 1D simulation with a constant-flux NeumannBC on the left side and a no-flux BC on the right side (which is the natural BC that you will obtain if you don't specify any BC). Make the domain the same size as one edge of the 2D domain you are doing for the full problem. Set the IC for the order parameter to one of the equilibrium values. Plot out the order parameter and look at what it's doing. Is is just piling up at the boundary? Or is it diffusing in to the remainder of the domain? If that works OK try with a 2-phase initial condition in 1D, where half the domain has one equilibrium value of the order parameter and the other equilibrium value.
